### ISSUE TITLE
TTY stopped working as expected in Docker version 20.10.17, build 100c701, API 1.42

### ISSUE THREAD

**Issue by kc-dot-io** [moby/moby#43799]
### Description

First, thanks to all the contributors for their hard work on this project, greatly appreciated. 👍 

We have a custom end user CLI that leverages [apocas/dockerode](https://github.com/apocas/dockerode) as a client to communicate with the Docker engine directly for the purpose of executing scripts inside of a container. These scripts typically have an interactive user experience in them and thus stdin/stdout and the container attach functionality is very important for TTY.

We've received some reports that this TTY has stopped working suddenly upon upgrading to`Docker version 20.10.17, build 100c701`. It appears that downgrading back to `Docker version 20.10.16, build aa7e414` fixes this immediately.

I found similar public reports from [JetBrains for their IDE](https://youtrack.jetbrains.com/issue/IDEA-297446) as well which seems to validate there is possibly a breaking change `20.10.17` for other docker clients which require TTY, as well.

The [`20.10.17` release notes](https://github.com/moby/moby/releases/tag/v20.10.17) didn't suggest any big changes but then I reviewed [the changes introduced between `20.10.16` & `20.10.17`](https://github.com/moby/moby/compare/v20.10.16...v20.10.17) and I came across [this API documentation change](https://github.com/moby/moby/commit/829951ec19c6762d358c0a6dae10c648bc515e4b) & [this change to the server that appears to specifically limit some Stdin, Stdout, Stderr functionality to v1.42 of the API](https://github.com/moby/moby/commit/ea6760138c35c607a0b51ad041051374d8304e68)?

I can see from the commit comments that @ndeloof is indicating that these APIs were documented but not supported prior to this version, however this (or similar) appears to be a breaking change for some custom clients such as ours unfortunately.

I'm looking for a bit more context to help me understand the impact of this change as it seems to very directly relate to the broken TTY problem being reported upstream. I'm still trying to break this down into a simple enough set of replication steps, but thought I might get this issue open in the meantime to see if anyone else is seeing this. Thanks for patience here.

### Reproduce

Apologies, I don't have simple replication steps yet (I'm working on it) without creating or using our custom client.

1. create a custom client using dockerode 
2. create the container, attach to the container
3. pipe stdin/stdout to support interactive TTY
4. stdin should be passed to container appropriately

_Please note_ I have tested this with the docker client using a simple `read` test in bash and it *does* work.

### Expected behavior

Expecting non breaking API for custom clients in `Docker Desktop 4.10.0` (`Docker version 20.10.17, build 100c701`), which require TTY when connecting directly to the Docker daemon (not using docker client).

### docker version

```bash
❯ docker version                                                                                        [15:57:17]
Client:
 Cloud integration: v1.0.24
 Version:           20.10.17
 API version:       1.41
 Go version:        go1.17.11
 Git commit:        100c701
 Built:             Mon Jun  6 23:04:45 2022
 OS/Arch:           darwin/amd64
 Context:           default
 Experimental:      true

Server: Docker Desktop 4.10.0 (82025)
 Engine:
  Version:          20.10.17
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.17.11
  Git commit:       a89b842
  Built:            Mon Jun  6 23:01:23 2022
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          1.6.6
  GitCommit:        10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1
 runc:
  Version:          1.1.2
  GitCommit:        v1.1.2-0-ga916309
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.8.2)
  compose: Docker Compose (Docker Inc., v2.6.1)
  extension: Manages Docker extensions (Docker Inc., v0.2.7)
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc., 0.6.0)
  scan: Docker Scan (Docker Inc., v0.17.0)

Server:
 Containers: 22
  Running: 0
  Paused: 0
  Stopped: 22
 Images: 30
 Server Version: 20.10.17
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1
 runc version: v1.1.2-0-ga916309
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.10.104-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 6
 Total Memory: 7.773GiB
 Name: docker-desktop
 ID: 4LKV:R7KC:SONZ:LMAU:Y37L:66SS:KWEN:6RXB:JJR7:2PTJ:FC77:26ZN
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 46
  Goroutines: 49
  System Time: 2022-07-12T22:57:32.145616762Z
  EventsListeners: 4
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: true
 Insecure Registries:
  hubproxy.docker.internal:5000
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by cpuguy83** [comment]
The specified change is not part of the 20.10 release. The API version v1.42 is also not in the 20.10 release.
Are you sure you are talking to a 20.10 engine when this issue occurs?

**Comment by thaJeztah** [comment]
@kc-dot-io do you know if this issue was reported only with users on docker desktop? Wondering if the issue is caused by a change in docker desktop

**Comment by kc-dot-io** [comment]
@cpuguy83 I could have it wrong so I will defer to you. I'm not 100% familiar with the dependency chain or release cycle, but I drilled down from the release notes on Docker Desktop 4.10.0 and felt pretty confident that this was the only change that seemed like it could have impated this. Ultimately the impact was just that I had to enable "hijack" option in the API but it does appear to me at least that it's a breaking change as our client has been stable for many many releases.

@thaJeztah I guess that's possible. I couldn't see anything in their release notes that seemed as relevant as what I linked to above though and I'm pretty confident that this is an issue with the engine / API opposed to any of the stuff around the edges as we have a pretty low level client integration directly with the engine itself but given I'm installing them via Docker Desktop it's certainly possible I missed something in the dependecy tree. 

Happy to anaswer any more questions if helpful.

**Comment by thaJeztah** [comment]
> we have a pretty low level client integration directly with the engine itself but given I'm installing them via Docker Desktop it's certainly possible I missed something in the dependecy tree.

By default, Docker Desktop uses a proxy in for the Engine's API, so when connecting to `/var/run/docker.sock`, that's a proxied version of the API (Desktop needs this proxy to rewrite some requests to adjust for paths being different within the VM and/or to set up filesystem mappings between the host and the VM etc), so it's possible there's been some changes there that caused a change in behaviour.

It's possible to skip the proxy by using `/var/run/docker.sock.raw` instead of `/var/run/docker.sock`, but I think the "raw" (non-proxied) socket is only available when bind-mounting into a container (it's mounted from within the VM).

I've never used [dockerode](https://github.com/apocas/dockerode)  myself; is there an easy way to reproduce the issue and/or do you have a way to run that reproducer from within a container that has `/var/run/docker.sock.raw` mounted ?


Something like;

```
docker run -it --rm -v /var/run/docker.sock.raw:/var/run/docker.sock <some-image-with-dockerode>
```

Of course, if you have the ability to try if it reproduces on a bare linux host with docker installed, that could also allow us to see if the regression is in Docker Desktop (not the engine).

**Comment by kc-dot-io** [comment]
Interesting. This is very helpful context. I had no idea Desktop proxied engine like this so it definitely could be something upstream with Desktop that I just couldn't find in the release notes, given this. Thanks for sharing that.

I did a bit more searching within dockerode and I actually came across [this thread](https://github.com/apocas/dockerode/issues/455) where there seems to be some very specific discussion from the point of view of the dockerode community on what seems like the same issue (sorry I missed this in my initial research). I think this might be helpful context as the examples they are sharing to replicate the behavior are almost the exact same as in our client and they also clearly call out a work around.

These examples could be tested with your suggested approach @thaJeztah, should you think its valid.

If you think it's still warranted, let me know and I can probably help set up the test case with dockerode.

I can't explain the lag between their report and this one. Our fork of dockerode is a bit behind, but not that far and it's pretty clear to me that 4.10.0 is the catalyst where this hijack option became a mandatory parameter for TTY (it wasn't before).
